_posts/2016-04-26-distributed-game-of-life-in-elixir.md

```diff
@@ -0,0 +1,1065 @@
+  use ExUnit.Case, async: true
+
+  test "alive cell with no neighbours dies" do
+    alive_cell = {1, 1}
+    alive_cells = [alive_cell]
+    refute GameOfLife.Cell.keep_alive?(alive_cells, alive_cell)
+  end
+
+  test "alive cell with 1 neighbour dies" do
+    alive_cell = {1, 1}
+    alive_cells = [alive_cell, {0, 0}]
+    refute GameOfLife.Cell.keep_alive?(alive_cells, alive_cell)
+  end
+end
+{% endhighlight %}
+
+`GameOfLife.Cell.keep_alive?/2` function needs to return false just to pass our tests so let's add more tests to cover other requirements.
+
+> Any live cell with more than three live neighbours dies, as if by over-population.
+
+{% highlight elixir %}
+# test/game_of_life/cell_test.exs
+test "alive cell with more than 3 neighbours dies" do
+  alive_cell = {1, 1}
+  alive_cells = [alive_cell, {0, 0}, {1, 0}, {2, 0}, {1, 0}]
+  refute GameOfLife.Cell.keep_alive?(alive_cells, alive_cell)
+end
+{% endhighlight %}
+
+> Any live cell with two or three live neighbours lives on to the next generation.
+
+{% highlight elixir %}
+# test/game_of_life/cell_test.exs
+test "alive cell with 2 neighbours lives" do
+  alive_cell = {1, 1}
+  alive_cells = [alive_cell, {0, 0}, {1, 0}]
+  assert GameOfLife.Cell.keep_alive?(alive_cells, alive_cell)
+end
+
+test "alive cell with 3 neighbours lives" do
+  alive_cell = {1, 1}
+  alive_cells = [alive_cell, {0, 0}, {1, 0}, {2, 1}]
+  assert GameOfLife.Cell.keep_alive?(alive_cells, alive_cell)
+end
+{% endhighlight %}
+
+Now we can implement our `GameOfLife.Cell.keep_alive?/2` function.
+
+{% highlight elixir %}
+# lib/game_of_life/cell.ex
+defmodule GameOfLife.Cell do
+  def keep_alive?(alive_cells, {x, y} = _alive_cell) do
+    case count_neighbours(alive_cells, x, y, 0) do
+      2 -> true
+      3 -> true
+      _ -> false
+    end
+  end
+
+  defp count_neighbours([head_cell | tail_cells], x, y, count) do
+    increment = case head_cell do
+      {hx, hy} when hx == x - 1 and hy == y - 1 -> 1
+      {hx, hy} when hx == x     and hy == y - 1 -> 1
+      {hx, hy} when hx == x + 1 and hy == y - 1 -> 1
+
+      {hx, hy} when hx == x - 1 and hy == y     -> 1
+      {hx, hy} when hx == x + 1 and hy == y     -> 1
+
+      {hx, hy} when hx == x - 1 and hy == y + 1 -> 1
+      {hx, hy} when hx == x     and hy == y + 1 -> 1
+      {hx, hy} when hx == x + 1 and hy == y + 1 -> 1
+
+      _not_neighbour -> 0
+    end
+    count_neighbours(tail_cells, x, y, count + increment)
+  end
+
+  defp count_neighbours([], _x, _y, count), do: count
+end
+{% endhighlight %}
+
+As you can see we implemented private function `count_neighbours/4` responsible for counting neighbours. It will be helpful later.
+
+There is one more requirement we forogot which is:
+
+> Any dead cell with exactly three live neighbours becomes a live cell, as if by reproduction.
+
+We are going to write a new function `GameOfLife.Cell.become_alive?/2` expecting coordinates of dead cell and returning if the dead cell should become alive or not.
```

Comment:
grammarly.com plugin shows me that there is one advanced issue in the sentence you proposed but I don't have pro plan to see what that is so I'm going to leave the first version.